staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+		return nil
+	}
+
+	// Retrieve all the other subnets. Doing a live list as opposed to a
+	// cache-based one (through the informer) prevents race conditions that can
+	// arise in case of multiple validators running.
+	allSubnets, err := v.netIfc.Subnets(k8scorev1api.NamespaceAll).List(k8smetav1.ListOptions{})
+	if err != nil && doNotRetryList(err) {
+		glog.Errorf("live list of all subnets against API server failed while validating %s: %s. There will be no retry because of the nature of the error", ss.NamespacedName, err.Error())
+		// This should never happen, no point in retrying.
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("live list of all subnets against API server failed: %s", err.Error())
+	}
+
+	// Look for conflicts with all the other subnets and record them in the
+	// rivals conflicts caches.
+	conflictsMsgs, conflictFound, err := v.recordConflicts(ss, allSubnets.Items)
+	if err != nil {
+		return err
+	}
+
+	if err := v.updateSubnetValidity(s, !conflictFound, conflictsMsgs); err != nil {
+		return fmt.Errorf("failed to write validation outcome into %s's status: %s", ss.NamespacedName, err.Error())
+	}
+
+	return nil
+}
+
+func (v *Validator) clearStaleRV(s k8stypes.NamespacedName) {
+	v.staleRVsMutex.Lock()
+	defer v.staleRVsMutex.Unlock()
+
+	delete(v.staleRVs, s)
+}
+
+func (v *Validator) clearConflictsCache(s k8stypes.NamespacedName) []k8stypes.NamespacedName {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	if c := v.conflicts[s]; c != nil {
+		delete(v.conflicts, s)
+		return c.rivals
+	}
+
+	return nil
+}
+
+func (v *Validator) subnetIsStale(s k8stypes.NamespacedName, rv string) bool {
+	v.staleRVsMutex.Lock()
+	defer v.staleRVsMutex.Unlock()
+
+	if v.staleRVs[s] == rv {
+		return true
+	}
+
+	delete(v.staleRVs, s)
+	return false
+}
+
+func (v *Validator) updateConflictsCache(s *subnet.Summary) []k8stypes.NamespacedName {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	c := v.conflicts[s.NamespacedName]
+	if c == nil {
+		c = &conflictsCache{
+			ownerSummary: s,
+			rivals:       make([]k8stypes.NamespacedName, 0),
+		}
+		v.conflicts[s.NamespacedName] = c
+		return nil
+	}
+
+	var oldRivals []k8stypes.NamespacedName
+	if s.VNI != c.ownerSummary.VNI || !s.Contains(c.ownerSummary) {
+		// The data that affects validation changed, return all rivals so that
+		// they can be re-validated again as the conflict might have disappeared.
+		oldRivals = c.rivals
+		c.rivals = make([]k8stypes.NamespacedName, 0)
+	}
+	c.ownerSummary = s
+
+	return oldRivals
+}
+
+func doNotRetryList(e error) bool {
+	return k8serrors.IsUnauthorized(e) ||
+		k8serrors.IsBadRequest(e) ||
+		k8serrors.IsForbidden(e) ||
+		k8serrors.IsNotAcceptable(e) ||
+		k8serrors.IsUnsupportedMediaType(e) ||
+		k8serrors.IsMethodNotSupported(e)
+}
+
+func (v *Validator) recordConflicts(candidate *subnet.Summary, potentialRivals []netv1a1.Subnet) ([]string, bool, error) {
+	var conflictsMsgs []string
```

Comment:
Might as well declare these in the function's return tuple.

Reply:
Fixed in 8ac9883.